course/api/views.py

```diff
@@ -64,6 +95,112 @@ def get_queryset(self):
+                        name = form.cleaned_data.get('name'),
+                        code = code,
+                        url = form.cleaned_data.get('course_url'),
+                    )
+                except:
+                    return Response('Course creation failed.',
+                                    status=status.HTTP_400_BAD_REQUEST)
+
+            formdata = form.cleaned_data
+            instance_name = formdata.get('instance_name')
+            if (not formdata.get('url')):
+                # if url is not given, compose it from instance name, removing special chars
+                formdata['url'] = ''.join(i for i in instance_name if i.isalnum())
+            
+            try:
+                new_instance: CourseInstance = CourseInstance.objects.create(
+                    instance_name = instance_name,
+                    url = formdata.get('url'),
+                    language = formdata.get('language'),
+                    starting_time = formdata.get('starting_time'),
+                    ending_time = formdata.get('ending_time'),
+                    visible_to_students = formdata.get('visible_to_students'),
+                    configure_url = formdata.get('configure_url'),
+                    course = course,
+                )
+                formdata['id'] = new_instance.id
+            except:
+                return Response('Course creation failed.',
+                        status=status.HTTP_400_BAD_REQUEST)
+
+            self.set_teachers(new_instance, data)
+
+        return Response(
+            {'status': 'success', 'data': formdata},
+            status=status.HTTP_201_CREATED
+        )
+
+    def update(self, request: HttpRequest, *args, **kwargs) -> Response:
+        if (not request.user.is_superuser):
+            return Response(
+                'Only admins can add new course instance.',
+                status=status.HTTP_403_FORBIDDEN
+            )
+
+        data = request.data
+        form = CourseInstanceForm(data)
+
+        if not form.is_valid():
+            return Response({'status': 'error', 'errors': form.errors}, status=status.HTTP_400_BAD_REQUEST)
+
+        formdata = form.cleaned_data
+        try:
+            course: CourseInstance = CourseInstance.objects.get(id=self.kwargs['course_id'])
+            formdata['id'] = course.id
+            course.instance_name = formdata.get('instance_name')
+            course.url = formdata.get('url')
+            course.language = formdata.get('language')
+            course.starting_time = formdata.get('starting_time')
+            course.ending_time = formdata.get('ending_time')
+            course.visible_to_students = formdata.get('visible_to_students')
+            course.configure_url = formdata.get('configure_url')
+            self.set_teachers(course, data)
+
+            course.save()
+
+        except CourseInstance.DoesNotExist:
+            return Response('Course instance does not exist.',
```

Comment:
Does DRF actually call this `update` method if the course id does not exist? It might check existence beforehand based on the ViewSet queryset, but I am not sure what DRF does.